### PR TITLE
Some big re-organizations to split 80 and 443 frontends + SNI support

### PR DESCRIPTION
**Pull request by faisyl**
While working on refactoring the code to not step on the ReqMode support, had to move some things around to support SNI and arbitrary TLS frontends. I needed to split 443 out so I could set that frontend to TCP mode so SNI will work, but that was a lot more complicated than I thought. Here's my take on the implementation. I'll write up the test cases for it and send you a pull request after you review this code. I'm new to golang and my apologies if the code isn't great. 

To use SNI mode routing, setting reqMode to sni does the job in combination with pathType (typically to 'req_ssl_sni -i -m beg') and servicePath to the matching SNI prefix.

### CODE REVIEW

File: actions/reconfigure.go [hunk 1]
```diff
@@ -153,7 +153,7 @@ func (m *Reconfigure) reloadFromRegistry(addresses []string, instanceName, mode
 		var data map[string]interface{}
 		json.Unmarshal(body, &data)
 		count = len(data)
-		for key, _ := range data {
+		for key := range data {
```
Comment (vfarcic) [inline actions/reconfigure.go]: `range` returns two values that depend on the collection it uses. When `range` is used with a map, the first value is the key and the second is value. If one of the return values is not used, it should be specified as `_`. I'm surprised this change even compiled.

File: haproxy.cfg [hunk 2]
```diff
@@ -26,7 +26,6 @@ defaults
 
 frontend dummy-fe
     bind *:80
-    bind *:443
```
Comment (vfarcic) [inline haproxy.cfg]: This file is used only when there are no services. HAProxy does not start without a valid config so I put this dummy file.

**Comment by vfarcic** [comment]
Overall, judging by reading the code (without running it), it looks good :).

Maybe I make a suggestion? Why not break changes into small chunks. For example, one change could be to move part of the .tmpl file into code (what you did with haproxy.tmpl). It's a small change which can be easily validated with a few tests, and send directly to a release. Continuous deployment approach. What do you think?

File: actions/reconfigure.go (hunk 1 above)
Comment (faisyl) [inline actions/reconfigure.go]: This was actually done by gofmt. Was using sublime and was formatting all the files I was changing. So instead of relying on that - I ran gofmt on the tree to make sure the formatting was clean.

File: haproxy.cfg (hunk 2 above)
Comment (faisyl) [inline haproxy.cfg]: Shall revert this change

**Comment by faisyl** [comment]
That makes sense. I'll keep the changes except the SNI related changes - since they are somewhat orthogonal. Do I need to add more test cases around this re-org? If so, some pointers on what to test would be useful.

**Comment by vfarcic** [comment]
I almost always do TDD. I write a test that defines code I'm about to write. I'll add tests if you have trouble with them.

**Comment by faisyl** [comment]
Yes please. I'd appreciate that. I can use this to understand how to write test cases for code I submit in the future. Shall I send you a new pull request with the SNI stuff pulled out? Or do you want to add the tests to my branch and we can merge it back into your tree after that's done?

**Comment by vfarcic** [comment]
I'm practicing continuous deployment and every commit to the master branch is converted into a public release if all tests pass. This process is executed only on the master branch. Without tests it is very risky to merge a PR.

The easiest would be to do a PR on a branch (any but master). That way I can merge it without any risk, add tests, and merge it again to master. Would that work for you?

**Comment by faisyl** [comment]
That would work perfectly. Could you please create a new branch on your repo I can send a PR against? Not sure how to consolidate the changes if it's against any of the existing stale branches.

**Comment by vfarcic** [comment]
Created `sni` branch :)
